*Autor Ismael Aguilera* - Trabajo de fin de grado - Detección de vehículos en videos de tráfico



## Instalar dependencias

In [14]:
# %%bash
!pip install -qr https://raw.githubusercontent.com/ultralytics/yolov5/master/requirements.txt  # install dependencies
!pip install ffmpeg
from google.colab import drive
drive.mount('/content/drive')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Generar información (subprograma)


In [15]:
#!pip install Pillow==9.0.0
import torch
import torchvision
import numpy as np
import torch
import matplotlib.pyplot as plt
import torchvision.transforms.functional as F
import torchvision.transforms as T
import cv2
import pandas as pd

#import PIL
#!pip install -U pillow
#!pip install pytesseract

def generarInformacion(rutaImg, rutaInfo):
  
  # Model
  model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)

  # Inference
  results = model(rutaImg)

  # Results
  results.print()
  results.save(rutaInfo)  # or .show()

  results.xyxy[0]  # img1 predictions (tensor)
  results.pandas().xyxy[0]  # img1 predictions (pandas)
  file1 = open(rutaInfo, "w")
  rutaInfoCsv = rutaInfo + ".csv"
  file1.write(str(results.pandas().xyxy[0]))
  file1.close()
  results.pandas().xyxy[0].to_csv(rutaInfoCsv)

## Pintar información (subprograma)

In [16]:
from IPython.core.display import Image

import cv2
import numpy as np
import csv
from csv import reader
import math
import pandas as pd
import cv2 as cv
#from tracker import *
!pip install norfair[video]
!pip install yolov5
import norfair
from norfair import Detection, Tracker, Video, Paths

import io
from base64 import b64encode
from IPython.display import HTML

conta = 0;

detections = []

def pintarInformacion2(rutaImg, rutaInfo, rutaPintada):

  # Creamos el object tracker
  #tracker = EuclideanDistTracker()

  #cap = cv2.VideoCapture(filename)

  # Creamos el object detector
  #object_detector = cv2.createBackgroundSubtractorMOG2(history=100, varThreshold=40)
      
  #Coger valores de csv
  with open(rutaInfo, newline='') as f:
    reader = csv.reader(f,delimiter=',')
    line = list(reader)

  with open(rutaInfo, "r") as fp:
    numlin = len(fp.readlines()) 

  frame = cv2.imread(rutaImg)


  # Hacemos la detección de objetos
  i = 1
  while i<numlin:
    x = int(float(line[i][1]))
    y = int(float(line[i][2]))
    w = int(float(line[i][3]))
    h = int(float(line[i][4]))
    con = int(float(line[i][5]))
    cx = int((x+w)/2)
    cy = int((y+h)/2)
    i = i + 1;
    cv2.rectangle(frame, (x,y), (w,h), (0,255,0), 2)    
      
    #detections.append([x, y, w, h])

  # Realizamos el object tracking
  #boxes_ids = tracker.update(detections)
  #for box_id in boxes_ids:
  #    x, y, w, h, id = box_id
  #    cv2.putText(frame, str(id), (x, y - 15), cv2.FONT_HERSHEY_PLAIN, 2, (255, 0, 0), 2)
      #cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 3)

  cv2.imwrite(rutaPintada,frame)

  #cap.release()
  cv2.destroyAllWindows()

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/



## Pintar tracking (subprograma)

In [17]:
#!wget "https://raw.githubusercontent.com/tryolabs/norfair/master/demos/yolov5/yolov5demo.py" -O yolov5demo.py
import shutil
shutil.copy('/content/drive/MyDrive/videostrafico/yolov5demo.py', "/content/")
def pintarTracking():
  !python yolov5demo.py sample.mp4 --img_size 1280 --conf_thres 0.4 --classes 0 1 2 3 4 5 6 7 8 --track_points 'bbox'

## Dividir en frames el video


In [ ]:
import cv2
import os
import subprocess
import shutil

filename = "/content/drive/MyDrive/videostrafico/videocarla1.mp4"
#filename = "/content/drive/MyDrive/videostrafico/cctv052x2004080519x01687.avi"

saveDirFilenames = "/content/drive/MyDrive/videostrafico/df/%02d.png"
print(saveDirFilenames)

fname = str(filename)
cap = cv2.VideoCapture(fname)
fps = cap.get(cv2.CAP_PROP_FPS)
amount_of_frames = cap.get(7)
print('Numero de frames en el video = ',amount_of_frames)
print('Extrayendo frames')
command = str('ffmpeg -i ' +'"'+ str(fname)+'"' + ' ' + '-q:v 1' + ' ' + '-start_number 0' + ' '+'"'+ str(saveDirFilenames)+'"')
print(command)
subprocess.call(command, shell=True)
print('Frames extraidos correctamente')


/content/drive/MyDrive/videostrafico/df/%02d.png


KeyboardInterrupt: ignored

## Generar información de cada frame

In [ ]:
amount_of_frames = 6051
rutaImgIni = "/content/drive/MyDrive/videostrafico/df/"
saveRutaInfoIni = "/content/drive/MyDrive/videostrafico/rutaInformacion/results"
cont = 0
while cont<amount_of_frames:
  if (cont < 10):
      rutaImg = rutaImgIni + "0" + str(cont) + ".png"
      #rutaImg = rutaImgIni + "in00000" + str(cont) + ".jpg"
  else:
      rutaImg = rutaImgIni + str(cont) + ".png"
  saveRutaInfo = saveRutaInfoIni + str(cont) + ".txt"
  print(rutaImg)
  print(saveRutaInfo)
  generarInformacion(rutaImg,saveRutaInfo)
  cont = cont + 1
cont = 0

Output hidden; open in https://colab.research.google.com to view.

## Pintar información de cada frame

In [ ]:
rutaImgIni = "/content/drive/MyDrive/videostrafico/df/"
RutaInfoIni = "/content/drive/MyDrive/videostrafico/rutaInformacion/results"
rutaImgPinIni = '/content/drive/MyDrive/videostrafico/imagenesPintadas/'
cont = 0
while cont<amount_of_frames:
  if (cont < 10):
      rutaImg = rutaImgIni + "0" + str(cont) + ".png"
      #rutaImg = rutaImgIni + "in00000" + str(cont) + ".jpg"
  else:
      rutaImg = rutaImgIni + str(cont) + ".png"
  RutaInfo = RutaInfoIni + str(cont) + ".txt.csv"
  rutaImgPin = rutaImgPinIni + str(cont) + "ip.png"
  print(rutaImg)
  print(RutaInfo)
  print(rutaImgPin)
  #pintarInformacion(rutaImg,RutaInfo, rutaImgPin)
  pintarInformacion2(rutaImg,RutaInfo, rutaImgPin)    
  cont = cont + 1
cont = 0

## Generar video en base a frames

In [ ]:
rutaImgPinIni = "/content/drive/MyDrive/videostrafico/df/"
cont = 1
img_array = []
rutaImg = rutaImgPinIni
while cont<amount_of_frames:
  if (cont < 10):
    rutaImg = rutaImgIni + "0" + str(cont) + ".png"
    #rutaImg = rutaImgIni + "in00000" + str(cont) + ".jpg"
  else:
      rutaImg = rutaImgIni + str(cont) + ".png"
  print(rutaImg)
  img = cv2.imread(rutaImg)
  height, width, layers = img.shape
  size = (width,height)
  img_array.append(img)

  cont = cont + 1

fourcc = cv2.VideoWriter_fourcc(*'mpeg')
out = cv2.VideoWriter()
out.open('/content/drive/MyDrive/videostrafico/videopintado/videoFrames.mp4',fourcc,10,size,True)

for i in range(len(img_array)):
  out.write(img_array[i])
out.release()

cont = 0

NameError: ignored

## Generar video con bounding boxes

In [ ]:
rutaImgPinIni = "/content/drive/MyDrive/videostrafico/imagenesPintadas/"
cont = 0
img_array = []
while cont<amount_of_frames:
  rutaImgPin = rutaImgPinIni + str(cont) + "ip.png"
  print(rutaImgPin)
  img = cv2.imread(rutaImgPin)
  height, width, layers = img.shape
  size = (width,height)
  img_array.append(img)

  cont = cont + 1

fourcc = cv2.VideoWriter_fourcc(*'mpeg')
out = cv2.VideoWriter()
out.open('/content/drive/MyDrive/videostrafico/videopintado/v1.mp4',fourcc,10,size,True)

for i in range(len(img_array)):
  out.write(img_array[i])
out.release()

cont = 0

## Pintar tracking

In [18]:
import shutil
shutil.copy('/content/drive/MyDrive/videostrafico/videopintado/videocarla1.mp4', "/content/sample.mp4")
pintarTracking()
shutil.copy("/content/sample_out.mp4", "/content/drive/MyDrive/videostrafico/videopintado/sampleVideoFrames.mp4")

Using cache found in /root/.cache/torch/hub/ultralytics_yolov5_master
YOLOv5 🚀 2022-10-3 Python-3.7.14 torch-1.12.1+cu113 CUDA:0 (Tesla T4, 15110MiB)

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
Adding AutoShape... 
 ...  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 11.08fps
Output video file saved to: ./sample_out.mp4
El máximo id es:  56
Coefficients: 
 [   -0.41144]
Coefficients: 
 [   0.026138]
Coefficients: 
 [    0.77895]
Coefficients: 
 [   0.054435]
Coefficients: 
 [   -0.67802]
Coefficients: 
 [     7.2447]
Coefficients: 
 [   -0.72678]
Coefficients: 
 [    0.86682]
Coefficients: 
 [   -0.36186]
Coefficients: 
 [  -0.099198]
Coefficients: 
 [   -0.19959]
Coefficients: 
 [   -0.82922]
Coefficients: 
 [    -0.8775]
Coefficients: 
 [     1.2138]
Coefficients: 
 [   -0.65035]
Coefficients: 
 [   -0.25868]
Coefficients: 
 [  -0.027128]
Coefficients: 
 [  -0.045827]
Coefficients: 
 [    -1.7441]
Coefficients: 
 [    -3.8728]
Coefficients: 

'/content/drive/MyDrive/videostrafico/videopintado/sampleVideoFrames.mp4'

## Generar video al juntar frames


In [ ]:
rutaImgPinIni = "/content/drive/MyDrive/videostrafico/imagenesPintadas/"
cont = 0
img_array = []
while cont<amount_of_frames:
  rutaImgPin = rutaImgPinIni + str(cont) + "ip.png"
  print(rutaImgPin)
  img = cv2.imread(rutaImgPin)
  height, width, layers = img.shape
  size = (width,height)
  img_array.append(img)

  cont = cont + 1

fourcc = cv2.VideoWriter_fourcc(*'mpeg')
out = cv2.VideoWriter()
out.open('/content/drive/MyDrive/videostrafico/videopintado/v1.mp4',fourcc,10,size,True)

for i in range(len(img_array)):
  out.write(img_array[i])
out.release()

cont = 0

## Model Description 

<img width="800" alt="YOLOv5 Model Comparison" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png">
&nbsp;

[YOLOv5](https://ultralytics.com/yolov5) 🚀 is a family of compound-scaled object detection models trained on the COCO dataset, and includes simple functionality for Test Time Augmentation (TTA), model ensembling, hyperparameter evolution, and export to ONNX, CoreML and TFLite.

|Model |size<br><sup>(pixels) |mAP<sup>val<br>0.5:0.95 |mAP<sup>test<br>0.5:0.95 |mAP<sup>val<br>0.5 |Speed<br><sup>V100 (ms) | |params<br><sup>(M) |FLOPS<br><sup>640 (B)
|---   |---  |---        |---         |---             |---                |---|---              |---
|[YOLOv5s6](https://github.com/ultralytics/yolov5/releases)   |1280 |43.3     |43.3     |61.9     |**4.3** | |12.7  |17.4
|[YOLOv5m6](https://github.com/ultralytics/yolov5/releases)   |1280 |50.5     |50.5     |68.7     |8.4     | |35.9  |52.4
|[YOLOv5l6](https://github.com/ultralytics/yolov5/releases)   |1280 |53.4     |53.4     |71.1     |12.3    | |77.2  |117.7
|[YOLOv5x6](https://github.com/ultralytics/yolov5/releases)   |1280 |**54.4** |**54.4** |**72.0** |22.4    | |141.8 |222.9
|[YOLOv5x6](https://github.com/ultralytics/yolov5/releases) TTA |1280 |**55.0** |**55.0** |**72.0** |70.8 | |-  |-

<details>
  <summary>Table Notes (click to expand)</summary>

  * AP<sup>test</sup> denotes COCO [test-dev2017](http://cocodataset.org/#upload) server results, all other AP results denote val2017 accuracy.
  * AP values are for single-model single-scale unless otherwise noted. **Reproduce mAP** by `python test.py --data coco.yaml --img 640 --conf 0.001 --iou 0.65`
  * Speed<sub>GPU</sub> averaged over 5000 COCO val2017 images using a GCP [n1-standard-16](https://cloud.google.com/compute/docs/machine-types#n1_standard_machine_types) V100 instance, and includes FP16 inference, postprocessing and NMS. **Reproduce speed** by `python test.py --data coco.yaml --img 640 --conf 0.25 --iou 0.45`
  * All checkpoints are trained to 300 epochs with default settings and hyperparameters (no autoaugmentation).
  * Test Time Augmentation ([TTA](https://github.com/ultralytics/yolov5/issues/303)) includes reflection and scale augmentation. **Reproduce TTA** by `python test.py --data coco.yaml --img 1536 --iou 0.7 --augment`

</details>

<p align="left"><img width="800" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_plot.png"></p>

<details>
  <summary>Figure Notes (click to expand)</summary>

  * GPU Speed measures end-to-end time per image averaged over 5000 COCO val2017 images using a V100 GPU with batch size 32, and includes image preprocessing, PyTorch FP16 inference, postprocessing and NMS.
  * EfficientDet data from [google/automl](https://github.com/google/automl) at batch size 8.
  * **Reproduce** by `python test.py --task study --data coco.yaml --iou 0.7 --weights yolov5s6.pt yolov5m6.pt yolov5l6.pt yolov5x6.pt`

</details>

## Load From PyTorch Hub


This example loads a pretrained **YOLOv5s** model and passes an image for inference. YOLOv5 accepts **URL**, **Filename**, **PIL**, **OpenCV**, **Numpy** and **PyTorch** inputs, and returns detections in **torch**, **pandas**, and **JSON** output formats. See our [YOLOv5 PyTorch Hub Tutorial](https://github.com/ultralytics/yolov5/issues/36) for details.

## Citation

[![DOI](https://zenodo.org/badge/264818686.svg)](https://zenodo.org/badge/latestdoi/264818686)


## Contact


**Issues should be raised directly in https://github.com/ultralytics/yolov5.** For business inquiries or professional support requests please visit [https://ultralytics.com](https://ultralytics.com) or email Glenn Jocher at [glenn.jocher@ultralytics.com](mailto:glenn.jocher@ultralytics.com).


&nbsp;